### PR TITLE
[CONTINT-3706] Increase the fake intake memory

### PR DESCRIPTION
**Pull request by L3n41c**
What does this PR do?
---------------------

Bump the memory of the fake intake to avoid OOM kill on the EKS and Kind scenarios.
Reduce the memory of the fake intake to save money on other scenarios.

Which scenarios this will impact?
-------------------

* all

Motivation
----------

I was looking at some flaky e2e tests and some of the failures are pointing towards the fake intake.

Ex.:

* [fake intake returns 504 error](https://app.datadoghq.com/ci/test-runs?query=test_level%3Atest%20%40test.service%3Adatadog-agent%20%40git.branch%3Amain%20%40test.suite%3A%22github.com%2FDataDog%2Fdatadog-agent%2Ftest%2Fnew-e2e%2Ftests%2Fcontainers%22%20%40test.status%3Afail&citest_explorer_sort=time%2Cdesc&currentTab=overview&eventStack=AgAAAY1P4T3aaWxAuwAAAAAAAAAYAAAAAEFZMVA0VDNhQUFCSXVaSXFhQ0xkTmNfTQAAACQAAAAAMDE4ZDRmZTMtMTFjMy00YzNjLTlmODItZTQxNjhhMmZmODg0&index=citest&mode=sliding&saved-view-id=2236559&testId=AgAAAY1P4T3aaWxAuwAAAAAAAAAYAAAAAEFZMVA0VDNhQUFCSXVaSXFhQ0xkTmNfTQAAACQAAAAAMDE4ZDRmZTMtMTFjMy00YzNjLTlmODItZTQxNjhhMmZmODg0&trace=AgAAAY1P4T3aaWxAuwAAAAAAAAAYAAAAAEFZMVA0VDNhQUFCSXVaSXFhQ0xkTmNfTQAAACQAAAAAMDE4ZDRmZTMtMTFjMy00YzNjLTlmODItZTQxNjhhMmZmODg0&start=1705919011102&end=1706523811102&paused=false):
  ```
        	Error:      	Received unexpected error:
        	            	Expected 200 got 504
        	Messages:   	Failed to query fake intake
  ```
* [fake intake returns 502 error](https://app.datadoghq.com/ci/test-runs?query=test_level%3Atest%20%40test.service%3Adatadog-agent%20%40git.branch%3Amain%20%40test.suite%3A%22github.com%2FDataDog%2Fdatadog-agent%2Ftest%2Fnew-e2e%2Ftests%2Fcontainers%22%20%40test.status%3Afail&citest_explorer_sort=time%2Cdesc&currentTab=overview&eventStack=AgAAAY1PzvYqHbuP1wAAAAAAAAAYAAAAAEFZMVB6dllxQUFDWGFZeld0TUhwYUp1NgAAACQAAAAAMDE4ZDRmZDUtNWViZC00MzcyLTlhYjctNjRmNzBjMDliMjk4&index=citest&mode=sliding&saved-view-id=2236559&testId=AgAAAY1PzvYqHbuP1wAAAAAAAAAYAAAAAEFZMVB6dllxQUFDWGFZeld0TUhwYUp1NgAAACQAAAAAMDE4ZDRmZDUtNWViZC00MzcyLTlhYjctNjRmNzBjMDliMjk4&trace=AgAAAY1PzvYqHbuP1wAAAAAAAAAYAAAAAEFZMVB6dllxQUFDWGFZeld0TUhwYUp1NgAAACQAAAAAMDE4ZDRmZDUtNWViZC00MzcyLTlhYjctNjRmNzBjMDliMjk4&start=1705919131102&end=1706523931102&paused=false)
  ```
        	Error:      	Received unexpected error:
        	            	Expected 200 got 502
        	Messages:   	Failed to query fake intake
  ```
* [fake intake misses data](https://app.datadoghq.com/ci/test-runs?query=test_level%3Atest%20%40test.service%3Adatadog-agent%20%40git.branch%3Amain%20%40test.suite%3A%22github.com%2FDataDog%2Fdatadog-agent%2Ftest%2Fnew-e2e%2Ftests%2Fcontainers%22%20%40test.status%3Afail&citest_explorer_sort=time%2Cdesc&currentTab=overview&eventStack=AgAAAY1P0MMaTZdOTwAAAAAAAAAYAAAAAEFZMVAwTU1hQUFEdWlCUkVQNk1BeG5GcwAAACQAAAAAMDE4ZDRmZDUtNWViZC00MzcyLTlhYjctNjRmNzBjMDliMjk4&index=citest&mode=sliding&saved-view-id=2236559&testId=AgAAAY1P0MMaTZdOTwAAAAAAAAAYAAAAAEFZMVAwTU1hQUFEdWlCUkVQNk1BeG5GcwAAACQAAAAAMDE4ZDRmZDUtNWViZC00MzcyLTlhYjctNjRmNzBjMDliMjk4&trace=AgAAAY1P0MMaTZdOTwAAAAAAAAAYAAAAAEFZMVAwTU1hQUFEdWlCUkVQNk1BeG5GcwAAACQAAAAAMDE4ZDRmZDUtNWViZC00MzcyLTlhYjctNjRmNzBjMDliMjk4&start=1705919191102&end=1706523991102&paused=false):
  ```
        	Error:      	Should NOT be empty, but was 
        	Messages:   	No SBOM for ghcr.io/datadog/apps-nginx-server yet
  ```

The data are showing a fake intake memory starvation followed by a restart:

* [First case](https://dddev.datadoghq.com/dashboard/qcp-brm-ysc/e2e-tests-containers-k8s?refresh_mode=paused&tpl_var_fake_intake_task_family%5B0%5D=ci-27290983-4670-kind-cluster-fakeintake-ecs&tpl_var_kube_cluster_name%5B0%5D=ci-27290983-4670-kind-cluster&view=spans&from_ts=1706441346106&to_ts=1706443060420&live=false):
  
image
* [Seconde case](https://dddev.datadoghq.com/dashboard/qcp-brm-ysc/e2e-tests-containers-k8s?refresh_mode=paused&tpl_var_fake_intake_task_family%5B0%5D=ci-27290998-4670-kind-cluster-fakeintake-ecs&tpl_var_kube_cluster_name%5B0%5D=ci-27290998-4670-kind-cluster&view=spans&from_ts=1706440152971&to_ts=1706441824091&live=false):
  
image

In both cases, we can see exactly the same symptoms.
* The memory consumption starts to plateau (because of `GOMEMLIMIT`) before the “Number of payloads collected” starts to plateau. This means that the garbage collector is asked to do cleanup before the fake intake starts to cleanup its payloads.
* This results in pressure that is visible on the CPU graph: the garbage collector consumes more and more CPU to try to keep the memory consumption below `GOMEMLIMIT`. Up to 1 core (the limit of the Fargate task).
* As all the CPU available to the Fargate task is consumed by the garbage collector, the fake intake response time degrades. It climbs up to more than 1 second.
* And the fake intake is eventually killed and restarted.

Additional Notes
----------------

The default CPU and memory limits used for the non-Kubernetes scenarios have been chosen based on the measures collected in [this notebook](https://dddev.datadoghq.com/notebook/7384273/fake-intake-memory-consumption).

### CODE REVIEW

File: scenarios/aws/fakeintake/fakeintake.go [hunk 1]
```diff
@@ -246,7 +246,7 @@ func fargateLinuxContainerDefinition(imageURL string, apiKeySSMParamName pulumi.
 		Environment: awsxEcs.TaskDefinitionKeyValuePairArray{
 			awsxEcs.TaskDefinitionKeyValuePairArgs{
 				Name:  pulumi.StringPtr("GOMEMLIMIT"),
-				Value: pulumi.StringPtr("3072MiB"),
+				Value: pulumi.StringPtr("5120MiB"),
```
[the remark this answers is not in the document]
Comment (pducolin) [reply scenarios/aws/fakeintake/fakeintake.go:249]: `GOMEMLIMIT` is expected to be~80% of the available memory as it is a soft limit

[the remark this answers is not in the document]
Comment (L3n41c) [reply scenarios/aws/fakeintake/fakeintake.go:249]: Indeed, that’s a “soft” memory limit.
The GO runtime does its best to remain below but cannot guarantee it.
Also, this is the limit for what the GO runtime allocates. We need to add a small margin for what’s not inside this enveloppe like the memory mapping of the binary for ex.

With the current limit set to `3072MiB`, we can already see the effective consumption going above:
image

[the remark this answers is not in the document]
Comment (vboulineau) [reply scenarios/aws/fakeintake/fakeintake.go]: Yes `GOMEMLIMIT` is a soft limit. Yet the 80% is probably a waste, especially as the limit grows bigger.
FWIW, measurements done in spiky traffic for `trace-agent` showed that `90%` provides the best target.

The fakeintake traffic is not super spiky and there's no memory usage outside the Go space, the binary is not huge, and IMHO by setting `GOMEMLIMIT` so low we're just not using what we have. So, something like `90% - 95%` or even a fixed overhead (which is probably closer to what actually happens of 200 or 300M) is probably enough.

[the remark this answers is not in the document]
Comment (L3n41c) [reply scenarios/aws/fakeintake/fakeintake.go]: We can definitively go for a fixed margin.
Here is the change: https://github.com/DataDog/test-infra-definitions/pull/586/files#diff-7677ec28dbaacd2bc5af4811b41a48f30563a9038d68e8ab5426a5aa2c01380cR63

I also checked the agent memory consumption in [this notebook](https://dddev.datadoghq.com/notebook/7384545/fake-intake-memory-consumption):
image

Should we take a fixed margin of 300 MiB (to account the fact that `GOMEMLIMIT` is a “soft” limit) + 400 MiB for the agent ?

**Review by pducolin — APPROVED** [review]

**Pull request merged**